### ISSUE TITLE
RocksDB thread safety in validateTransactionInputsExpensive

### ISSUE THREAD

**Issue by zpalmtree** [turtlecoin/turtlecoin#1022]
There appears to be an issue where multi-threading in [`validateTransactionInputsExpensive`](https://github.com/turtlecoin/turtlecoin/blob/26b6eac05372c2afa38cfeeacc310590ea5d0426/src/cryptonotecore/ValidateTransaction.cpp#L477) causes crashes. I have seen a number of crash logs similar to the following:

```
#0  0x00007f3ff0670f2e in malloc () from /usr/lib/libc.so.6
#1  0x000056291bc3f9da in operator new(unsigned long) ()
#2  0x000056291b848602 in rocksdb::IterKey::EnlargeBuffer(unsigned long) ()
#3  0x000056291b874889 in rocksdb::IterKey::TrimAppend(unsigned long, char const*, unsigned long) ()
#4  0x000056291b95e1b8 in bool rocksdb::DataBlockIter::ParseNextDataKey<rocksdb::DecodeEntry>(char const*) ()
#5  0x000056291b95d065 in rocksdb::DataBlockIter::Seek(rocksdb::Slice const&) ()
#6  0x000056291b8db942 in rocksdb::BlockBasedTable::Get(rocksdb::ReadOptions const&, rocksdb::Slice const&, rocksdb::GetContext*, rocksdb::SliceTransform const*, bool) ()
#7  0x000056291b87361e in rocksdb::TableCache::Get(rocksdb::ReadOptions const&, rocksdb::InternalKeyComparator const&, rocksdb::FileMetaData const&, rocksdb::Slice const&, rocksdb::GetContext*, rocksdb::SliceTransform const*, rocksdb::HistogramImpl*, bool, int) ()
#8  0x000056291b87bfe2 in rocksdb::Version::Get(rocksdb::ReadOptions const&, rocksdb::LookupKey const&, rocksdb::PinnableSlice*, rocksdb::Status*, rocksdb::MergeContext*, unsigned long*, bool*, bool*, unsigned long*, rocksdb::ReadCallback*, bool*, bool) ()
#9  0x000056291b7f8f32 in rocksdb::DBImpl::MultiGet(rocksdb::ReadOptions const&, std::vector<rocksdb::ColumnFamilyHandle*, std::allocator<rocksdb::ColumnFamilyHandle*> > const&, std::vector<rocksdb::Slice, std::allocator<rocksdb::Slice> > const&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >*) ()
#10 0x000056291b80751d in rocksdb::DB::MultiGet(rocksdb::ReadOptions const&, std::vector<rocksdb::Slice, std::allocator<rocksdb::Slice> > const&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >*) ()
#11 0x000056291b75233f in CryptoNote::RocksDBWrapper::read(CryptoNote::IReadBatch&) ()
#12 0x000056291b704c61 in CryptoNote::DatabaseBlockchainCache::checkIfSpent(Crypto::KeyImage const&, unsigned int) const ()
#13 0x000056291b78a68a in CryptoNote::BlockchainCache::checkIfSpent(Crypto::KeyImage const&, unsigned int) const ()
#14 0x000056291b78a68a in CryptoNote::BlockchainCache::checkIfSpent(Crypto::KeyImage const&, unsigned int) const ()
#15 0x000056291b78a68a in CryptoNote::BlockchainCache::checkIfSpent(Crypto::KeyImage const&, unsigned int) const ()
#16 0x000056291b780a9a in ValidateTransaction::validateTransactionInputsExpensive()::$_0::operator()() const ()
#17 0x000056291b78088d in std::_Function_handler<bool (), ValidateTransaction::validateTransactionInputsExpensive()::$_0>::_M_invoke(std::_Any_data const&) ()
#18 0x000056291b6c6e5e in std::function<bool ()>::operator()() const ()
#19 0x000056291b6c5228 in Utilities::ThreadPool<bool>::waitForJob() ()
#20 0x000056291b6ca017 in void std::__invoke_impl<void, void (Utilities::ThreadPool<bool>::*)(), Utilities::ThreadPool<bool>*>(std::__invoke_memfun_deref, void (Utilities::ThreadPool<bool>::*&&)(), Utilities::ThreadPool<bool>*&&) ()
#21 0x000056291b6c9f22 in std::__invoke_result<void (Utilities::ThreadPool<bool>::*)(), Utilities::ThreadPool<bool>*>::type std::__invoke<void (Utilities::ThreadPool<bool>::*)(), Utilities::ThreadPool<bool>*>(void (Utilities::ThreadPool<bool>::*&&)(), Utilities::ThreadPool<bool>*&&) ()
#22 0x000056291b6c9ee5 in void std::thread::_Invoker<std::tuple<void (Utilities::ThreadPool<bool>::*)(), Utilities::ThreadPool<bool>*> >::_M_invoke<0ul, 1ul>(std::_Index_tuple<0ul, 1ul>) ()
#23 0x000056291b6c9e95 in std::thread::_Invoker<std::tuple<void (Utilities::ThreadPool<bool>::*)(), Utilities::ThreadPool<bool>*> >::operator()() ()
#24 0x000056291b6c9c5e in std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (Utilities::ThreadPool<bool>::*)(), Utilities::ThreadPool<bool>*> > >::_M_run() ()
#25 0x000056291bcc4ff4 in execute_native_thread_routine ()
#26 0x00007f3ff0c6f4cf in start_thread () from /usr/lib/libpthread.so.0
#27 0x00007f3ff06e52d3 in clone () from /usr/lib/libc.so.6
```

Likely, the fix is to use a mutex to synchronize access. However, where to put the mutex to avoid negating the benefit of threading the input validation is key.

Workaround is to launch with `--transaction-validation-threads 1`

**Comment by brandonlehmann** [comment]
May be fixed in #1032. Awaiting verification from others.

**Comment by brandonlehmann** [comment]
Considering closed via #1032 that was merged and released.
